generator/booking_quality_check/booking_quality_check.cpp

```diff
@@ -84,109 +113,177 @@ struct Emitter : public EmitterBase
-                  << endl;
-        outStream << "# " << PrintBuilder(fb) << endl;
-        outStream << "# " << hotel << endl;
-        outStream << "# URL: https://www.openstreetmap.org/?mlat=" << hotel.lat
-                  << "&mlon=" << hotel.lon << "#map=18/" << hotel.lat << "/" << hotel.lon << endl;
-      }
-      if (!bookingIndexes.empty())
-        outStream << endl << endl;
-    }
+  info.m_intermediateDir = my::GetDirectory(FLAGS_factors);
 
-    if (FLAGS_sample_data.empty())
-    {
-      cout << outStream.str();
-    }
-    else
-    {
-      ofstream file(FLAGS_sample_data);
-      if (file.is_open())
-      {
-        file << outStream.str();
-      }
-      else
-      {
-        LOG(LERROR, ("Can't output into", FLAGS_sample_data));
-        return false;
-      }
-    }
+  // Set other info params here.
 
-    return true;
+  return info;
+}
+
+struct SampleItem
+{
+  enum MatchStatus {Uninitialized, Yes, No};
+
+  SampleItem() = default;
+
+  SampleItem(osm::Id const & osmId, uint32_t const bookingId, MatchStatus const match = Uninitialized)
+   : m_osmId(osmId)
+   , m_bookingId(bookingId)
+   , m_match(match)
+  {
   }
 
-  BookingDataset m_bookingDataset;
-  vector<FeatureBuilder1> m_features;
+  osm::Id m_osmId;
+  uint32_t m_bookingId = BookingDataset::kInvalidHotelIndex;
+
+  MatchStatus m_match = Uninitialized;
 };
 
-unique_ptr<Emitter> GetEmitter(feature::GenerateInfo const & info)
+SampleItem::MatchStatus ReadMatchStatus(string const & str)
 {
-  LOG_SHORT(LINFO, ("Booking data:", FLAGS_booking_data));
-  return make_unique<Emitter>(info);
+  if (str == "Yes")
+    return SampleItem::Yes;
+
+  if (str == "No")
+    return SampleItem::No;
+
+  if (str == "Uninitialized")
+    return SampleItem::Uninitialized;
+
+  MYTHROW(ParseError, ("Can't make SampleItem::MatchStatus from string:", str));
 }
 
-feature::GenerateInfo GetGenerateInfo()
+SampleItem ReadSampleItem(string const & str)
 {
-  feature::GenerateInfo info;
-  info.m_bookingDatafileName = FLAGS_booking_data;
-  info.m_osmFileName = FLAGS_osm_file_name;
-  info.SetNodeStorageType("map");
-  info.SetOsmFileType("o5m");
+  SampleItem item;
 
-  info.m_intermediateDir =  my::GetDirectory(FLAGS_sample_data);
+  auto const parts = strings::Tokenize(str, "\t");
+  CHECK_EQUAL(parts.size(), 3, ("Cant't make SampleItem from string:", str,
+                                "due to wrong number of fields."));
 
-  // ...
+  item.m_osmId = ReadDebuggedPrintedOsmId(parts[0]);
+  if (!strings::to_uint(parts[1], item.m_bookingId))
+    MYTHROW(ParseError, ("Can't make uint32 from string:", parts[1]));
+  item.m_match = ReadMatchStatus(parts[2]);
 
-  return info;
+  return item;
+}
+
+vector<SampleItem> ReadSample(istream & ist)
+{
+  vector<SampleItem> result;
+
+  size_t lineNumber = 1;
+  try
+  {
+    for (string line; getline(ist, line); ++lineNumber)
+    {
+      result.emplace_back(ReadSampleItem(line));
+    }
+  }
+  catch (ParseError const & e)
```

Comment:
The reason is that you intensively use exceptions here, and no real reason to catch them here - this is the philosophy of unix cl tools - exit early in case of errors.